File: paleo_workbench/ui/modelview/reconcile.py
"""QTreeWidget/QListWidget 按键差分同步（消灭 clear+rebuild）。

模式取自两处已验证实现：``workstation/explorer._reconcile_children``
（标准项树键控差分）与 ``stratigraphy_correlation_page._sync_well_list``
（万井级列表增量复用）。V11 把它抽成通用工具，供仍保留 widget 形态
的小中型表面（图层树、文档列表、参考层清单等，01-ui-audit D2 ⑮）使用。

契约：

* 调用方给出**目标键序列**（有序、稳定业务键）与两个回调：
  ``make_item(key) -> item``（新建）和 ``update_item(item, key)``（刷新）；
* 返回 ``dict[key, item]``；已存在且键序不变的项对象**原样保留**
  （展开态/选择/滚动/自定义数据不丢）；
* 移除消失键、追加新增键、按目标顺序重排（仅当顺序变化时 take/insert）；
* 明确不做的：分页、过滤、排序——这些属于模型层职责。
"""
from __future__ import annotations

from collections.abc import Callable, Sequence

from PySide6.QtCore import Qt
from PySide6.QtWidgets import QListWidget, QListWidgetItem, QTreeWidget, QTreeWidgetItem

_KEY_ROLE = Qt.ItemDataRole.UserRole + 0x10D6  # D666 → “差分”专用段，避免与页面其它 UserRole 混用


def _item_key(item: QTreeWidgetItem | QListWidgetItem) -> str | None:
    return item.data(0, _KEY_ROLE) if isinstance(item, QTreeWidgetItem) else item.data(_KEY_ROLE)


def _set_item_key(item: QTreeWidgetItem | QListWidgetItem, key: str) -> None:
    if isinstance(item, QTreeWidgetItem):
        item.setData(0, _KEY_ROLE, key)
    else:
        item.setData(_KEY_ROLE, key)


def reconcile_widget_items(
    widget: QTreeWidget | QListWidget,
    keys: Sequence[str],
    *,
    make_item: Callable[[str], QTreeWidgetItem | QListWidgetItem],
    update_item: Callable[[QTreeWidgetItem | QListWidgetItem, str], None],
    parent_item: QTreeWidgetItem | None = None,
) -> dict[str, QTreeWidgetItem | QListWidgetItem]:
    """把 widget（或某父项）的子项集合同步为目标键序列。

    ``parent_item`` 仅对 QTreeWidget 有意义（None = 顶层）。
    """
    if isinstance(widget, QTreeWidget):
        container = parent_item if parent_item is not None else widget.invisibleRootItem()
        existing: list[QTreeWidgetItem] = [container.child(i) for i in range(container.childCount())]
    else:
        existing = [widget.item(i) for i in range(widget.count())]

    by_key: dict[str, QTreeWidgetItem | QListWidgetItem] = {}
    for item in existing:
        key = _item_key(item)
        if key is not None and key not in by_key:
            by_key[key] = item

    target_keys = [k for k in keys if k is not None]
    # 重复键去重（保序）：重复键会让同键两项都存活于差分之外（评审 P2-10）。
    seen: set[str] = set()
    deduped: list[str] = []
    for key in target_keys:
        if key not in seen:
            seen.add(key)
            deduped.append(key)
    target_keys = deduped
    wanted = set(target_keys)

    # 1) 移除消失键
    for item in existing:
        key = _item_key(item)
        if key is None or key not in wanted:
            if isinstance(widget, QTreeWidget):
                container.removeChild(item)  # type: ignore[union-attr]
            else:
                widget.takeItem(widget.row(item))
            by_key.pop(key, None)

    # 2) 新增缺失键（新项先入容器，再交调用方刷新内容）
    for key in target_keys:
        if key not in by_key:
            item = make_item(key)
            _set_item_key(item, key)
            if isinstance(widget, QTreeWidget):
                if parent_item is not None:
                    parent_item.addChild(item)
                else:
                    widget.addTopLevelItem(item)
            else:
                widget.addItem(item)
            by_key[key] = item

    # 3) 刷新内容（键序未变时零结构操作）
    for key in target_keys:
        update_item(by_key[key], key)

    # 4) 顺序对齐（只在失序时 take/insert，保选择与展开态）
    current_keys = []
    if isinstance(widget, QTreeWidget):
        current_keys = [
            _item_key(container.child(i)) for i in range(container.childCount())
        ]
    else:
        current_keys = [_item_key(widget.item(i)) for i in range(widget.count())]
    if current_keys != target_keys:
        for position, key in enumerate(target_keys):
            item = by_key[key]
            if isinstance(widget, QTreeWidget):
                row = _tree_row_of(container, item)  # type: ignore[union-attr]
                if row != position:
                    child = container.takeChild(row)  # type: ignore[union-attr]
                    container.insertChild(position, child)  # type: ignore[union-attr]
            else:
                row = widget.row(item)  # type: ignore[call-arg]
                if row != position:
                    widget.takeItem(row)
                    widget.insertItem(position, item)

    return by_key


def _tree_row_of(container: QTreeWidgetItem, item: QTreeWidgetItem) -> int:
    for i in range(container.childCount()):
        if container.child(i) is item:
            return i
    return -1


def item_key(item: QTreeWidgetItem | QListWidgetItem) -> str | None:
    """读取 :func:`reconcile_widget_items` 写入的键（页面侧查询用）。"""
    return _item_key(item)


__all__ = ["reconcile_widget_items", "item_key"]

File: libs/ui_widgets/src/reconcile.cpp
#include "pwb/ui_widgets/reconcile.hpp"

#include <set>

namespace pwb::ui_widgets {
namespace {

// Dedupe preserving first occurrence (review P2-10: duplicated keys
// would otherwise let same-key extra items live outside the differ).
std::vector<QString> deduped_keys(const std::vector<QString>& keys) {
    std::set<QString> seen;
    std::vector<QString> out;
    out.reserve(keys.size());
    for (const QString& key : keys) {
        // Python `if k is not None` — empty string is a valid key there
        // (str key space); keep it here too.
        if (seen.insert(key).second) {
            out.push_back(key);
        }
    }
    return out;
}

int tree_row_of(QTreeWidgetItem* container, QTreeWidgetItem* item) {
    for (int i = 0; i < container->childCount(); ++i) {
        if (container->child(i) == item) return i;
    }
    return -1;
}

QString tree_item_key(QTreeWidgetItem* item) {
    return item->data(0, kReconcileKeyRole).toString();
}

bool tree_item_has_key(QTreeWidgetItem* item) {
    const QVariant v = item->data(0, kReconcileKeyRole);
    return v.isValid() && !v.isNull();
}

QString list_item_key(QListWidgetItem* item) {
    return item->data(kReconcileKeyRole).toString();
}

bool list_item_has_key(QListWidgetItem* item) {
    const QVariant v = item->data(kReconcileKeyRole);
    return v.isValid() && !v.isNull();
}

}  // namespace

QString item_key(QTreeWidgetItem* item) {
    return item == nullptr ? QString() : tree_item_key(item);
}

QString item_key(QListWidgetItem* item) {
    return item == nullptr ? QString() : list_item_key(item);
}

std::map<QString, QTreeWidgetItem*> reconcile_widget_items(
    QTreeWidget* widget,
    const std::vector<QString>& keys,
    const std::function<QTreeWidgetItem*(const QString&)>& make_item,
    const std::function<void(QTreeWidgetItem*, const QString&)>& update_item,
    QTreeWidgetItem* parent_item) {
    QTreeWidgetItem* container = parent_item != nullptr
                                     ? parent_item
                                     : widget->invisibleRootItem();
    std::vector<QTreeWidgetItem*> existing;
    existing.reserve(size_t(container->childCount()));
    for (int i = 0; i < container->childCount(); ++i) {
        existing.push_back(container->child(i));
    }

    std::map<QString, QTreeWidgetItem*> by_key;
    for (QTreeWidgetItem* item : existing) {
        // Python `key is not None` — keyless items never enter by_key
        // (they get removed in step 1 below).
        if (tree_item_has_key(item)) {
            const QString key = tree_item_key(item);
            if (by_key.find(key) == by_key.end()) {
                by_key[key] = item;
            }
        }
    }

    const std::vector<QString> target_keys = deduped_keys(keys);
    const std::set<QString> wanted(target_keys.begin(), target_keys.end());

    // 1) Remove vanished keys. Python removes only items whose key is
    //    None or not in wanted — items with NO stored key are removed
    //    too (key None not in wanted). Items with keys still wanted but
    //    losing the by_key race (duplicates) survive — frozen quirk.
    //    Removed items lose all references in Python (GC frees them) ->
    //    delete here.
    for (QTreeWidgetItem* item : existing) {
        const QString key = tree_item_key(item);
        if (!tree_item_has_key(item) ||
            wanted.find(key) == wanted.end()) {
            container->removeChild(item);
            delete item;
            by_key.erase(key);
        }
    }

    // 2) Add missing keys (item enters the container first, then the
    //    caller refreshes its content).
    for (const QString& key : target_keys) {
        if (by_key.find(key) == by_key.end()) {
            QTreeWidgetItem* item = make_item(key);
            item->setData(0, kReconcileKeyRole, key);
            if (parent_item != nullptr) {
                parent_item->addChild(item);
            } else {
                widget->addTopLevelItem(item);
            }
            by_key[key] = item;
        }
    }

    // 3) Refresh content (zero structural ops when the key order is
    //    unchanged).
    for (const QString& key : target_keys) {
        update_item(by_key[key], key);
    }

    // 4) Order alignment (take/insert only when out of order — keeps
    //    selection and expand state).
    std::vector<QString> current_keys;
    current_keys.reserve(size_t(container->childCount()));
    for (int i = 0; i < container->childCount(); ++i) {
        current_keys.push_back(tree_item_key(container->child(i)));
    }
    if (current_keys != target_keys) {
        for (int position = 0; position < int(target_keys.size());
             ++position) {
            QTreeWidgetItem* item = by_key[target_keys[size_t(position)]];
            const int row = tree_row_of(container, item);
            if (row != position && row >= 0) {
                QTreeWidgetItem* child = container->takeChild(row);
                container->insertChild(position, child);
            }
        }
    }
    return by_key;
}

std::map<QString, QListWidgetItem*> reconcile_widget_items(
    QListWidget* widget,
    const std::vector<QString>& keys,
    const std::function<QListWidgetItem*(const QString&)>& make_item,
    const std::function<void(QListWidgetItem*, const QString&)>& update_item) {
    std::vector<QListWidgetItem*> existing;
    existing.reserve(size_t(widget->count()));
    for (int i = 0; i < widget->count(); ++i) {
        existing.push_back(widget->item(i));
    }

    std::map<QString, QListWidgetItem*> by_key;
    for (QListWidgetItem* item : existing) {
        if (list_item_has_key(item)) {
            const QString key = list_item_key(item);
            if (by_key.find(key) == by_key.end()) {
                by_key[key] = item;
            }
        }
    }

    const std::vector<QString> target_keys = deduped_keys(keys);
    const std::set<QString> wanted(target_keys.begin(), target_keys.end());

    // 1) Remove vanished keys (same rules as the tree path).
    for (QListWidgetItem* item : existing) {
        const QString key = list_item_key(item);
        if (!list_item_has_key(item) ||
            wanted.find(key) == wanted.end()) {
            delete widget->takeItem(widget->row(item));
            by_key.erase(key);
        }
    }

    // 2) Add missing keys.
    for (const QString& key : target_keys) {
        if (by_key.find(key) == by_key.end()) {
            QListWidgetItem* item = make_item(key);
            item->setData(kReconcileKeyRole, key);
            widget->addItem(item);
            by_key[key] = item;
        }
    }

    // 3) Refresh content.
    for (const QString& key : target_keys) {
        update_item(by_key[key], key);
    }

    // 4) Order alignment.
    std::vector<QString> current_keys;
    current_keys.reserve(size_t(widget->count()));
    for (int i = 0; i < widget->count(); ++i) {
        current_keys.push_back(list_item_key(widget->item(i)));
    }
    if (current_keys != target_keys) {
        for (int position = 0; position < int(target_keys.size());
             ++position) {
            QListWidgetItem* item = by_key[target_keys[size_t(position)]];
            const int row = widget->row(item);
            if (row != position && row >= 0) {
                widget->takeItem(row);
                widget->insertItem(position, item);
            }
        }
    }
    return by_key;
}

}  // namespace pwb::ui_widgets
